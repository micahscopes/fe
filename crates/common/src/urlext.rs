use camino::Utf8PathBuf;
use url::Url;

#[derive(Debug)]
pub enum UrlExtError {
    DirectoryRangeError,
    AsDirectoryError,
}

pub trait UrlExt {
    fn parent(&self) -> Option<Url>;
    fn directory(&self) -> Option<Url>;
    fn join_directory(&self, path: &Utf8PathBuf) -> Result<Url, ()>;
}

impl UrlExt for Url {
    fn directory(&self) -> Option<Url> {
        if self.cannot_be_a_base() {
            return None;
        };
        let mut url = self.clone();

        if url.path().ends_with('/') {
            return Some(url);
        }

        if let Ok(mut segments) = url.path_segments_mut() {
            segments.pop();
            segments.push("");
        }

        Some(url)
    }

    fn parent(&self) -> Option<Url> {
        let directory = self.directory()?;

        if *self != directory {
            // If we're not already at a directory, return the directory
            Some(directory)
        } else if self.path() == "/" {
            return None;
        } else {
            // We're already at a directory, go up one level
            let mut parent = self.clone();
            if let Ok(mut segments) = parent.path_segments_mut() {
                segments.pop();
                segments.pop();
                segments.push(""); // Ensure trailing slash
            }
            return Some(parent);
        }
    }

    fn join_directory(&self, path: &Utf8PathBuf) -> Result<Url, ()> {
        Ok(if path.as_str().ends_with("/") {
            self.join(path.as_str()).map_err(|_| ())?
        } else {
            let mut path = path.clone();
            path.push("");
            self.join(path.as_str()).map_err(|_| ())?
        })
    }
}

pub fn canonical_url(path: &Utf8PathBuf) -> Result<Url, ()> {
    Ok(
        Url::from_directory_path(path.canonicalize_utf8().map_err(|_| ())?.as_str())
            .map_err(|_| ())?,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_directory_basic() {
        let url = Url::parse("https://example.com/foo/bar/baz").unwrap();
        let directory = url.directory();
        assert!(directory.is_some());
        assert_eq!(directory.unwrap().as_str(), "https://example.com/foo/bar/");

        let url = Url::parse("https://example.com/foo/bar/baz/").unwrap();
        let directory = url.directory();
        assert!(directory.is_some());
        assert_eq!(
            directory.unwrap().as_str(),
            "https://example.com/foo/bar/baz/"
        );
    }

    #[test]
    fn test_parent_basic() {
        let url = Url::parse("https://example.com/foo/bar").unwrap();
        let parent = url.parent();
        assert!(parent.is_some());
        assert_eq!(parent.unwrap().as_str(), "https://example.com/foo/");

        let url = Url::parse("https://example.com/foo/").unwrap();
        let parent = url.parent();
        assert!(parent.is_some());
        assert_eq!(parent.unwrap().as_str(), "https://example.com/");

        let url = Url::parse("https://example.com/").unwrap();
        let parent = url.parent();
        assert!(
            parent.is_none(),
            "Parent should be `None` but instead we got {:?}",
            parent
        );
    }

    #[test]
    fn test_file_url_parent_behavior() {
        // Test file:// URL paths
        let file_url = Url::parse("file:///foo/bar/baz.txt").unwrap();

        // Test directory() behavior
        let dir = file_url.directory();
        assert!(dir.is_some());
        assert_eq!(dir.unwrap().as_str(), "file:///foo/bar/");

        // Test parent() behavior - from file to directory
        let parent = file_url.parent();
        assert!(parent.is_some());
        assert_eq!(parent.unwrap().as_str(), "file:///foo/bar/");

        // Test parent of directory
        let dir_url = Url::parse("file:///foo/bar/").unwrap();
        let parent = dir_url.parent();
        assert!(parent.is_some());
        let parent_url = parent.unwrap();
        assert_eq!(parent_url.as_str(), "file:///foo/");

        // Test parent of parent
        let parent_of_parent = parent_url.parent();
        assert!(parent_of_parent.is_some());
        assert_eq!(parent_of_parent.unwrap().as_str(), "file:///");

        // Test parent of root
        let root_url = Url::parse("file:///").unwrap();
        let root_parent = root_url.parent();
        assert!(root_parent.is_none(), "Root URL should have no parent");
    }
}
