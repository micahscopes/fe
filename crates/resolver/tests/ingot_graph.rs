use std::str::FromStr;

use camino::Utf8PathBuf;
use common::urlext::UrlExt;
use fe_resolver::{ingot::basic_ingot_graph_resolver, Resolver};
use test_utils::snap_test;
use url::Url;

#[test]
fn ingot_graph_resolution() {
    let mut resolver = basic_ingot_graph_resolver();
    let base_url = Url::from_directory_path(
        Utf8PathBuf::from_str(".")
            .unwrap()
            .canonicalize_utf8()
            .unwrap(),
    )
    .unwrap();
    let ingot_a_url = base_url
        .join("fixtures/ingots/A")
        .unwrap()
        .directory()
        .unwrap();
    println!("{:?}", ingot_a_url);
    let ingot_graph = resolver.transient_resolve(&ingot_a_url).unwrap();
    snap_test!(
        format!("{:#?}\n{:#?}", ingot_graph, resolver.take_diagnostics()),
        "../../../fixtures/ingots"
    );
}
