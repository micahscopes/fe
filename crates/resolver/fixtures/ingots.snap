---
source: crates/resolver/tests/ingot_graph.rs
expression: "format!(\"{:#?}\\n{:#?}\", ingot_graph, resolver.take_diagnostics())"
---
Graph {
    Ty: "Directed",
    node_count: 8,
    edge_count: 8,
    edges: (0, 1), (1, 2), (2, 3), (3, 4), (2, 5), (1, 6), (0, 7), (7, 1),
    node weights: {
        0: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/A/",
            query: None,
            fragment: None,
        },
        1: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/C/",
            query: None,
            fragment: None,
        },
        2: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/E/",
            query: None,
            fragment: None,
        },
        3: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/G/",
            query: None,
            fragment: None,
        },
        4: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/H/",
            query: None,
            fragment: None,
        },
        5: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/F/",
            query: None,
            fragment: None,
        },
        6: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/D/",
            query: None,
            fragment: None,
        },
        7: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/home/grantwuerker/workshop/fe/crates/resolver/fixtures/ingots/B/",
            query: None,
            fragment: None,
        },
    },
    edge weights: {
        0: (
            "C",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
        1: (
            "E",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
        2: (
            "G",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
        3: (
            "H",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
        4: (
            "F",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
        5: (
            "D",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
        6: (
            "B",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
        7: (
            "C",
            IngotArguments {
                name: None,
                version: None,
            },
        ),
    },
}
[]
